src/Service/S3/src/S3Client.php

```diff
@@ -652,6 +652,27 @@ public function uploadPart($input): UploadPartOutput
         return new UploadPartOutput($response);
     }
 
+    protected function getEndpoint(string $uri, array $query, ?string $region): string
+    {
+        $uriParts = \explode('/', $uri, 3);
+        $bucket = $uriParts[1] ?? '';
+        $bucketLen = \strlen($bucket);
+        $configuration = $this->getConfiguration();
+
+        /** @psalm-suppress PossiblyNullArgument */
```

Comment:
beause `$configuration->get('endpoint')` can return `?string`, but will never return`null` because the `endpoint` property has a default value.

Let me thinking that we have a coupling between `S3` and `Core`, and `S3` knows how behave the internal implementation of `Core/Configuration`.... 
Shouldn't we provide a `getString` method ? /cc @Nyholm

Reply:
I don't use `\assert(is_string($bucket));` because I don't want executing code at runtime to hide a false positive.

Reply:
Im fine with the current code. The 6 options that have a default value will never be null when using `beause $configuration->get()`. That is something that is very hard for psalm to know, that is why the annotation is fine.

Reply:
@Nyholm it's easy to tell psalm this. Can be done when https://github.com/vimeo/psalm/issues/3419 is fixed.